lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -134,6 +144,95 @@ class ProgramLoweringState {
   DenseMap<Operation *, ComponentLoweringState> compStates;
 };
 
+//===----------------------------------------------------------------------===//
+// Partial lowering infrastructure
+//===----------------------------------------------------------------------===//
+
+/// Base class for partial lowering passes. A partial lowering pass
+/// modifies the root operation in place, but does not replace the root
+/// operation.
+/// The RewritePatternType template parameter allows for using both
+/// OpRewritePattern (default) or OpInterfaceRewritePattern.
+template <class OpType,
+          template <class> class RewritePatternType = OpRewritePattern>
+class PartialLoweringPattern : public RewritePatternType<OpType> {
+public:
+  using RewritePatternType<OpType>::RewritePatternType;
+  PartialLoweringPattern(MLIRContext *ctx, LogicalResult &resRef)
+      : RewritePatternType<OpType>(ctx), partialPatternRes(resRef) {}
+
+  LogicalResult matchAndRewrite(OpType op,
+                                PatternRewriter &rewriter) const override {
+    rewriter.updateRootInPlace(
+        op, [&] { partialPatternRes = partiallyLower(op, rewriter); });
+    return partialPatternRes;
+  }
+
+  virtual LogicalResult partiallyLower(OpType op,
+                                       PatternRewriter &rewriter) const = 0;
+
+private:
+  LogicalResult &partialPatternRes;
+};
+
+//===----------------------------------------------------------------------===//
+// Partial lowering patterns
+//===----------------------------------------------------------------------===//
+
+/// FuncOpPartialLoweringPatterns are patterns which intend to match on FuncOps
+/// and then perform their own walking of the IR. FuncOpPartialLoweringPatterns
+/// have direct access to the ComponentLoweringState for the corresponding
+/// component of the matched FuncOp.
+class FuncOpPartialLoweringPattern
+    : public PartialLoweringPattern<mlir::FuncOp> {
+public:
+  FuncOpPartialLoweringPattern(MLIRContext *context, LogicalResult &resRef,
+                               FuncMapping &_funcMap, ProgramLoweringState &pls)
+      : PartialLoweringPattern(context, resRef), funcMap(_funcMap), pls(pls) {}
+
+  LogicalResult partiallyLower(mlir::FuncOp funcOp,
+                               PatternRewriter &rewriter) const override final {
+    // Initialize the component op references if a calyx::ComponentOp has been
+    // created for the matched funcOp.
+    auto it = funcMap.find(funcOp);
```

Comment:
Likewise, this may be obvious, but I'm not following why a ComponentOp wouldn't be initialized for a given FuncOp. Is this a corner case thing where some FuncOps don't meet specific criteria (e.g. are too "simple")?

Reply:
This will only be the case for the pattern that is going to do the `FuncOp` to `ComponentOp` conversion.

I apologize for various things looking a bit weird/out of place in these initial PRs. I think (hope) things will start to make more sense once the following commits start to appear...